jumanji/environments/swarms/search_and_rescue/env.py

```diff
@@ -0,0 +1,390 @@
+                "Search & rescue multi-agent environment:",
+                f" - num searchers: {self.generator.num_searchers}",
+                f" - num targets: {self.generator.num_targets}",
+                f" - search vision range: {self.searcher_vision_range}",
+                f" - target contact range: {self.target_contact_range}",
+                f" - num vision: {self.num_vision}",
+                f" - agent radius: {self.agent_radius}",
+                f" - max steps: {self.max_steps},"
+                f" - target dynamics: {self._target_dynamics.__class__.__name__}",
+                f" - generator: {self.generator.__class__.__name__}",
+            ]
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Initialise searcher positions and velocities, and target positions.
+
+        Args:
+            key: Random key used to reset the environment.
+
+        Returns:
+            state: Initial environment state.
+            timestep: TimeStep with individual search agent views.
+        """
+        state = self.generator(key, self.searcher_params)
+        timestep = restart(observation=self._state_to_observation(state))
+        return state, timestep
+
+    def step(self, state: State, actions: chex.Array) -> Tuple[State, TimeStep[Observation]]:
+        """Environment update.
+
+        Update searcher velocities and consequently their positions,
+        mark found targets, and generate rewards and local observations.
+
+        Args:
+            state: Environment state.
+            actions: Arrays of searcher steering actions.
+
+        Returns:
+            state: Updated searcher and target positions and velocities.
+            timestep: Transition timestep with individual agent local observations.
+        """
+        searchers = update_state(state.key, self.searcher_params, state.searchers, actions)
+        key, target_key = jax.random.split(state.key, num=2)
+        # Ensure target positions are wrapped
+        target_pos = self._target_dynamics(target_key, state.targets.pos) % 1.0
+        # Grant searchers rewards if in range and not already detected
+        # spatial maps the has_found_target function over all pair of targets and
+        # searchers within range of each other and sums rewards per agent.
+        rewards = spatial(
+            utils.has_found_target,
+            reduction=jnp.add,
+            default=0.0,
+            i_range=self.target_contact_range,
+        )(
+            key,
+            self.searcher_params.view_angle,
+            searchers,
+            state.targets,
+            pos=searchers.pos,
+            pos_b=target_pos,
+        )
+        # Mark targets as found if with contact range and view angle of a searcher
+        # spatial maps the has_been_found function over all pair of targets and
+        # searchers within range of each other
+        targets_found = spatial(
+            utils.has_been_found,
```

Comment:
With these two function it seems like you're doing the same computation twice, I'd recommend checking which targets are found this step and then passing that to your reward function (which you could then move out of utils and put in here because it should be quite simple)

Reply:
I think this need to be done twice (or at least I could not really see a way round) given there is a kind of asymmetry given a searcher can detect multiple targets inside one step (at least as it's implemented), and so we can't know the size of the values that would need storing ahead of time (for example we might try to store the indices of the searcher linked to each target to create rewards).

One workaround could be to only allow one detection, or have some fixed number of detections each step?

Reply:
Very happy if you had another solution though, spent a while thinking about this 😆, this is one of the pain-points with JAX and simulations like this where interactions can variable size.

Reply:
Ye it can be quite tricky, usually the solution is to use a mask. 

So in this case the mask would be `targets_found & ~targets.found[:, jnp.newaxis]` (I'm not 100% sure of if the newaxis should be dim 0 or 1) I think this should output a matrix of `(num_searchers, num_targets)` which would indicate which targets are found by which searchers in the current step. Then your reward for each searcher is just the sum along it's dimension.

Reply:
Oh yeah this is an option, my concern with this approach though is the need to carry flags for each target through the map reduce for each agent, so in total I think it would scale like `9 * n_searchers * n_targets` (the nine comes from the number of regions checked around each agent in parallel). For larger numbers this could cause some memory issues.

Reply:
Though maybe mitigated if JAX has proper bit arrays?

Reply:
This may be overthinking/overoptimising though 😂 in scales of 10-100s of searchers to 100-1000s of targets this is in the Mb range (rather than Gb)

Reply:
I'm confused though how does the current implementation scale differently? (This is likely due to my lack of understanding around spatial, I'm picturing a nested `vmap`?)

Reply:
Yeah it is effectively, so the space is subdivided into a grid (dependent on interaction range), then each agent gathers results from each grid cell, then those are reduced to a result for a single agent. So just adds an extra _fixed_ factor for results for each cell, so not really different from what you are saying.
Was thinking about the memory v computation trade-off, but in practise it's probably not that bad to just use the extra memory (for some reason had the memory usage being worse from a previous project)!